### PR TITLE
Test to check that the cloud ingress operator is installed.

### PR DESCRIPTION
**Pull request by aliceh**
In this test we do a basic check on the cloud-ingress-operator deployment, verify that it is deployed with the correct number of replicas.

### CODE REVIEW

File: pkg/e2e/operators/cloudingress/installed.go [hunk 1]
```diff
@@ -1,31 +1,24 @@
...
 
 	"github.com/openshift/osde2e/pkg/common/helper"
-
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // tests
+
+func checkDeployment(h *helper.H, namespace string, name string, defaultDesiredReplicas int32)
+
```
Comment (yithian) [inline pkg/e2e/operators/cloudingress/installed.go:12]: what's going on here? this is go syntax i'm not familiar with

Comment (aliceh) [reply pkg/e2e/operators/cloudingress/installed.go:12]: That's an error on my part.

File: pkg/e2e/operators/cloudingress/installed.go [hunk 2]
```diff
@@ -2,32 +2,98 @@ package cloudingress
...
+	interval := 5
 
-// utils
+	// convert time.Duration type
+	timeoutDuration := time.Duration(viper.GetFloat64(config.Tests.PollingTimeout)) * time.Minute
+	intervalDuration := time.Duration(interval) * time.Second
+
+	start := time.Now()
+
+Loop:
```
Comment (yithian) [inline pkg/e2e/operators/cloudingress/installed.go:75]: I'm not familiar with this keyword - what's it do?

**Review by yithian — CHANGES_REQUESTED** [review]
for what it's worth, i ran this against a cluster in stage and it returned a failure:

```
• Failure [1805.242 seconds]
[Suite: informing] CloudIngressOperator
/home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:22
  deployment
  /home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:35
    should have all desired replicas ready [It]
    /home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:41

    failed fetching deployment
    Unexpected error:
        <*errors.errorString | 0xc000b940c0>: {
            s: "Failed to get cloud-ingress-operator Deployment before timeout",
        }
        Failed to get cloud-ingress-operator Deployment before timeout
    occurred

    /home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:43
```

File: pkg/e2e/operators/cloudingress/installed.go (hunk 2 above)
Comment (aliceh) [reply pkg/e2e/operators/cloudingress/installed.go:74]: It sets a reference point from which you can measure time elapsed. For example,  start := time.Now()  *something happens in between* end := time.Now()  time_elapsed:=start.Sub(end) More info [here](https://golang.org/pkg/time/) .

Comment (yithian) [reply pkg/e2e/operators/cloudingress/installed.go:73]: Sorry, I tried to put this comment on the `Loop:` line.

I didn't realize that loops could be labeled and targeted by a `break` statement like this. Neat!

File: pkg/e2e/operators/cloudingress/installed.go [hunk 3]
```diff
@@ -2,32 +2,96 @@ package cloudingress
...
-		}
+			readyReplicas := deployment.Status.ReadyReplicas
+			desiredReplicas := deployment.Status.Replicas
+
+			// The desired replicas should match the default installed replica count
+			Expect(desiredReplicas).To(BeNumerically("==", defaultDesiredReplicas), "The deployment desired replicas should not drift from the default 1.")
+
+			// Desired replica count should match ready replica count
+			Expect(readyReplicas).To(BeNumerically("==", desiredReplicas), "All desired replicas should be ready.")
+		}, float64(viper.GetFloat64(config.Tests.PollingTimeout)))
```
Comment (yithian) [inline pkg/e2e/operators/cloudingress/installed.go:51]: Do you foresee other files in the osde2e/pkg/e2e/operators/cloudingress package will need access to `checkDeployment()` ?

If not, it probably doesn't need to be a separate function.

File: pkg/e2e/operators/cloudingress/installed.go [hunk 4]
```diff
@@ -2,32 +2,96 @@ package cloudingress
...
-			Name string
-		}{}
+func checkDeployment(h *helper.H, namespace string, name string, defaultDesiredReplicas int32) {
+	// Check that the operator deployment exists in the operator namespace
+	ginkgo.Context("deployment", func() {
+		ginkgo.It("should exist", func() {
+			deployment, err := pollDeployment(h, namespace, name)
+			Expect(err).ToNot(HaveOccurred(), "failed fetching deployment")
+			Expect(deployment).NotTo(BeNil(), "deployment is nil")
+		}, float64(viper.GetFloat64(config.Tests.PollingTimeout)))
```
Comment (yithian) [inline pkg/e2e/operators/cloudingress/installed.go:38]:
```suggestion
		}, float64(viper.GetFloat64(config.Tests.PollingTimeout)))
		
```

let's put a newline here to make the tests more clearly separate

File: pkg/e2e/operators/cloudingress/installed.go (hunk 3 above)
Comment (aliceh) [reply pkg/e2e/operators/cloudingress/installed.go]: I agree with you.

Comment (yithian) [reply pkg/e2e/operators/cloudingress/installed.go]: Cool - please go ahead and put the body of `checkDeployment()` directly into the spot where it's called, above. It looks like it's the only call for it, so there's no need to DRY it up.

**Comment by yithian** [comment]
Tested this again today:

```
• Failure [1864.555 seconds]
[Suite: informing] CloudIngressOperator
/home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:22
  deployment
  /home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:33
    should exist [It]
    /home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:34

    failed fetching deployment
    Unexpected error:
        <*errors.errorString | 0xc000c20b30>: {
            s: "Failed to get cloud-ingress-operator Deployment before timeout",
        }
        Failed to get cloud-ingress-operator Deployment before timeout
    occurred

    /home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:36
...
• Failure [1803.957 seconds]
[Suite: informing] CloudIngressOperator
/home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:22
  deployment
  /home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:33
    should have all desired replicas ready [It]
    /home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:40

    failed fetching deployment
    Unexpected error:
        <*errors.errorString | 0xc000cb2dc0>: {
            s: "Failed to get cloud-ingress-operator Deployment before timeout",
        }
        Failed to get cloud-ingress-operator Deployment before timeout
    occurred

    /home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:42
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
JUnit report was created: /tmp/196574609/install/junit_r9i4w.xml


Summarizing 2 Failures:

[Fail] [Suite: informing] CloudIngressOperator deployment [It] should exist 
/home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:36

[Fail] [Suite: informing] CloudIngressOperator deployment [It] should have all desired replicas ready 
/home/achvatal/code/go/src/github.com/openshift/osde2e/pkg/e2e/operators/cloudingress/installed.go:42
```

**Comment by yithian** [comment]
tests passed for me now

**Pull request merged**